File: src/csharp/Intel/Iced/Intel/Internal/DataReader.cs
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

#if INSTR_INFO || DECODER || GAS || INTEL || MASM || NASM || FAST_FMT
using System;

namespace Iced.Intel.Internal {
#if HAS_SPAN
	ref struct DataReader {
		readonly ReadOnlySpan<byte> data;
#else
	struct DataReader {
		readonly byte[] data;
#endif
		readonly char[] stringData;
		int index;

		public int Index {
			readonly get => index;
			set => index = value;
		}

		public readonly bool CanRead => (uint)index < (uint)data.Length;

#if HAS_SPAN
		public DataReader(ReadOnlySpan<byte> data)
#else
		public DataReader(byte[] data)
#endif
			: this(data, 0) {
		}

#if HAS_SPAN
		public DataReader(ReadOnlySpan<byte> data, int maxStringLength) {
#else
		public DataReader(byte[] data, int maxStringLength) {
#endif
			this.data = data;
			stringData = maxStringLength == 0 ? Array2.Empty<char>() : new char[maxStringLength];
			index = 0;
		}

		public byte ReadByte() => data[index++];

		public uint ReadCompressedUInt32() {
			uint result = 0;
			for (int shift = 0; shift < 32; shift += 7) {
				uint b = ReadByte();
				if ((b & 0x80) == 0)
					return result | (b << shift);
				result |= (b & 0x7F) << shift;
			}
			throw new InvalidOperationException();
		}

		public string ReadAsciiString() {
			int length = ReadByte();
			var stringData = this.stringData;
			for (int i = 0; i < length; i++)
				stringData[i] = (char)ReadByte();
			return new string(stringData, 0, length);
		}
	}
}
#endif

File: src/java/iced-x86/src/main/java/com/github/icedland/iced/x86/internal/DataReader.java
// SPDX-License-Identifier: MIT
// Copyright (C) 2018-present iced project and contributors

package com.github.icedland.iced.x86.internal;

/**
 * DO NOT USE: INTERNAL API
 *
 * @deprecated Not part of the public API
 */
@Deprecated
public final class DataReader {
	private final byte[] data;
	private final char[] stringData;
	private int index;

	/** DO NOT USE: INTERNAL API */
	public DataReader(byte[] data) {
		this(data, 0);
	}

	/** DO NOT USE: INTERNAL API */
	public DataReader(byte[] data, int maxStringLength) {
		this.data = data;
		stringData = new char[maxStringLength];
		index = 0;
	}

	public boolean canRead() {
		return Integer.compareUnsigned(index, data.length) < 0;
	}

	/** DO NOT USE: INTERNAL API */
	public int getIndex() {
		return index;
	}

	/** DO NOT USE: INTERNAL API */
	public void setIndex(int index) {
		this.index = index;
	}

	/** DO NOT USE: INTERNAL API */
	public int readByte() {
		return data[index++] & 0xFF;
	}

	/** DO NOT USE: INTERNAL API */
	public int readCompressedUInt32() {
		int result = 0;
		for (int shift = 0; shift < 32; shift += 7) {
			int b = readByte();
			if ((b & 0x80) == 0)
				return result | (b << shift);
			result |= (b & 0x7F) << shift;
		}
		throw new UnsupportedOperationException();
	}

	/** DO NOT USE: INTERNAL API */
	public String readAsciiString() {
		int length = readByte();
		char[] stringData = this.stringData;
		for (int i = 0; i < length; i++)
			stringData[i] = (char)readByte();
		return new String(stringData, 0, length);
	}
}
